File: mcp-code-intelligence-kotlin/src/main/kotlin/com/codeintel/graph/CallGraphService.kt
/** Call Graph Service — BFS traversal for callers/callees. KSA-173. */
package com.codeintel.graph

import com.codeintel.graph.models.*
import java.sql.Connection
import java.util.ArrayDeque

class CallGraphService(
    private val conn: Connection,
    private val resolver: SymbolResolver,
) {

    /** Find all callers of a symbol with transitive depth. */
    fun findCallers(
        symbolName: String,
        depth: Int = 1,
        limit: Int = 20,
        fileFilter: String? = null,
        kindFilter: String = "calls",
    ): CallGraphResponse {
        val startTime = System.currentTimeMillis()
        val clampedDepth = depth.coerceIn(1, 5)

        val resolved = resolver.resolve(symbolName)
        if (resolved.isEmpty()) return symbolNotFound(symbolName)

        val results = mutableListOf<CallGraphItem>()
        val visited = mutableSetOf<Int>()
        val queue = ArrayDeque<Pair<String, Int>>()

        resolved.forEach { queue.add(it.name to 0) }

        while (queue.isNotEmpty() && results.size < limit) {
            val (current, currentDepth) = queue.poll()
            if (currentDepth >= clampedDepth) continue

            val callers = findCallersDb(current, kindFilter, limit - results.size)
            for (caller in callers) {
                if (caller.id in visited) continue
                visited.add(caller.id)

                val item = CallGraphItem(
                    symbol = caller.name,
                    qualifiedName = if (caller.parameters != null) "${caller.parameters}.${caller.name}" else caller.name,
                    kind = caller.kind,
                    filePath = caller.filePath,
                    definitionLine = caller.defLine,
                    callSiteLine = caller.callLine,
                    depthLevel = currentDepth + 1,
                    parameters = caller.parameters,
                    isAsync = caller.isAsync,
                )

                if (fileFilter != null && !matchFilter(item.filePath, fileFilter)) continue
                results.add(item)

                if (currentDepth + 1 < clampedDepth) {
                    queue.add(caller.name to currentDepth + 1)
                }
            }
        }

        val elapsed = System.currentTimeMillis() - startTime
        return CallGraphResponse(
            symbol = symbolName,
            resolvedTo = resolved.map { ResolvedTo(it.id, it.filePath, it.line, it.kind) },
            results = results,
            metadata = CallGraphMetadata(results.size, clampedDepth, results.size >= limit, elapsed),
        )
    }

    /** Find all callees of a symbol with transitive depth. */
    fun findCallees(
        symbolName: String,
        depth: Int = 1,
        limit: Int = 20,
        fileFilter: String? = null,
        includeExternal: Boolean = true,
    ): CallGraphResponse {
        val startTime = System.currentTimeMillis()
        val clampedDepth = depth.coerceIn(1, 5)

        val resolved = resolver.resolve(symbolName)
        if (resolved.isEmpty()) return symbolNotFound(symbolName)

        val results = mutableListOf<CallGraphItem>()
        val visited = mutableSetOf<String>()
        val queue = ArrayDeque<Pair<Int, Int>>()

        resolved.forEach { queue.add(it.id to 0) }

        while (queue.isNotEmpty() && results.size < limit) {
            val (symbolId, currentDepth) = queue.poll()
            if (currentDepth >= clampedDepth) continue

            val callees = findCalleesDb(symbolId, "calls", limit - results.size)
            for (callee in callees) {
                val key = "${callee.name}:${callee.callLine}"
                if (key in visited) continue
                visited.add(key)

                if (!includeExternal && callee.filePath == null) continue

                val item = CallGraphItem(
                    symbol = callee.name,
                    qualifiedName = callee.name,
                    kind = callee.kind ?: "unknown",
                    filePath = callee.filePath ?: "(external)",
                    definitionLine = callee.defLine ?: 0,
                    callSiteLine = callee.callLine,
                    depthLevel = currentDepth + 1,
                )

                if (fileFilter != null && item.filePath != "(external)" && !matchFilter(item.filePath, fileFilter)) continue
                results.add(item)

                if (callee.filePath != null && currentDepth + 1 < clampedDepth) {
                    val calleeResolved = resolver.resolve(callee.name)
                    calleeResolved.firstOrNull { it.filePath == callee.filePath }?.let {
                        queue.add(it.id to currentDepth + 1)
                    }
                }
            }
        }

        val elapsed = System.currentTimeMillis() - startTime
        return CallGraphResponse(
            symbol = symbolName,
            resolvedTo = resolved.map { ResolvedTo(it.id, it.filePath, it.line, it.kind) },
            results = results,
            metadata = CallGraphMetadata(results.size, clampedDepth, results.size >= limit, elapsed),
        )
    }

    private fun findCallersDb(symbolName: String, kind: String, limit: Int): List<CallerRow> {
        val stmt = conn.prepareStatement("""
            SELECT s.name, s.kind, s.file_path, s.start_line, r.line,
                   s.parent_symbol, s.visibility, s.id
            FROM relationships r
            JOIN symbols s ON s.id = r.source_symbol_id
            WHERE r.target_symbol = ? AND r.kind = ?
            ORDER BY s.file_path, r.line LIMIT ?
        """.trimIndent())
        stmt.setString(1, symbolName)
        stmt.setString(2, kind)
        stmt.setInt(3, limit)
        val rs = stmt.executeQuery()
        val rows = mutableListOf<CallerRow>()
        while (rs.next()) {
            rows.add(CallerRow(
                name = rs.getString(1),
                kind = rs.getString(2),
                filePath = rs.getString(3),
                defLine = rs.getInt(4),
                callLine = rs.getInt(5),
                parameters = rs.getString(6),
                isAsync = rs.getInt(7) == 1,
                id = rs.getInt(8),
            ))
        }
        return rows
    }

    private fun findCalleesDb(symbolId: Int, kind: String, limit: Int): List<CalleeRow> {
        val stmt = conn.prepareStatement("""
            SELECT r.target_symbol, r.line, r.metadata, ts.kind, ts.file_path, ts.start_line
            FROM relationships r
            LEFT JOIN symbols ts ON ts.id = r.target_symbol_id
            WHERE r.source_symbol_id = ? AND r.kind = ?
            ORDER BY r.line LIMIT ?
        """.trimIndent())
        stmt.setInt(1, symbolId)
        stmt.setString(2, kind)
        stmt.setInt(3, limit)
        val rs = stmt.executeQuery()
        val rows = mutableListOf<CalleeRow>()
        while (rs.next()) {
            rows.add(CalleeRow(
                name = rs.getString(1),
                callLine = rs.getInt(2),
                kind = rs.getString(4),
                filePath = rs.getString(5),
                defLine = rs.getInt(6).takeIf { !rs.wasNull() },
            ))
        }
        return rows
    }

    private fun symbolNotFound(symbolName: String) = CallGraphResponse(symbol = symbolName)

    private fun matchFilter(filePath: String, filter: String): Boolean {
        if ("*" in filter) {
            val regex = Regex("^" + Regex.escape(filter).replace("\\*", ".*") + "$")
            return regex.matches(filePath)
        }
        return filter in filePath
    }
}

private data class CallerRow(
    val name: String, val kind: String, val filePath: String,
    val defLine: Int, val callLine: Int, val parameters: String?,
    val isAsync: Boolean, val id: Int,
)

private data class CalleeRow(
    val name: String, val callLine: Int,
    val kind: String?, val filePath: String?, val defLine: Int?,
)

File: backend/mcp-code-intelligence-python/src/mcp_code_intel/graph/call_graph_service.py
"""Call Graph Service — BFS traversal for callers/callees. KSA-179."""

from __future__ import annotations

import re
import sqlite3
import time
from collections import deque

from .models import CallGraphItem, CallGraphMetadata, CallGraphResponse, ResolvedTo
from .symbol_resolver import SymbolResolver


class CallGraphService:
    """Provides transitive call graph analysis with depth control."""

    def __init__(self, conn: sqlite3.Connection, resolver: SymbolResolver) -> None:
        self._conn = conn
        self._resolver = resolver

    def find_callers(
        self,
        symbol_name: str,
        depth: int = 1,
        limit: int = 20,
        file_filter: str | None = None,
        kind_filter: str = "calls",
    ) -> CallGraphResponse:
        """Find all callers of a symbol with transitive depth."""
        start_time = time.time()
        clamped_depth = min(max(depth, 1), 5)

        resolved = self._resolver.resolve(symbol_name)
        if not resolved:
            return self._symbol_not_found(symbol_name)

        results: list[CallGraphItem] = []
        visited: set[int] = set()
        queue: deque[tuple[str, int]] = deque()

        for sym in resolved:
            queue.append((sym.name, 0))

        while queue and len(results) < limit:
            current_name, current_depth = queue.popleft()
            if current_depth >= clamped_depth:
                continue

            callers = self._find_callers_db(current_name, kind_filter, limit - len(results))

            for caller in callers:
                caller_id = caller["id"]
                if caller_id in visited:
                    continue
                visited.add(caller_id)

                item = CallGraphItem(
                    symbol=caller["name"],
                    qualified_name=f"{caller['parameters']}.{caller['name']}" if caller["parameters"] else caller["name"],
                    kind=caller["kind"],
                    file_path=caller["file_path"],
                    definition_line=caller["def_line"],
                    call_site_line=caller["call_line"],
                    depth_level=current_depth + 1,
                    parameters=caller["parameters"],
                    is_async=caller["is_async"] == 1,
                )

                if file_filter and not self._match_filter(item.file_path, file_filter):
                    continue

                results.append(item)

                if current_depth + 1 < clamped_depth:
                    queue.append((caller["name"], current_depth + 1))

        elapsed_ms = int((time.time() - start_time) * 1000)
        return CallGraphResponse(
            symbol=symbol_name,
            resolved_to=[ResolvedTo(id=s.id, file=s.file_path, line=s.line, kind=s.kind) for s in resolved],
            results=results,
            metadata=CallGraphMetadata(
                total_count=len(results),
                depth_searched=clamped_depth,
                truncated=len(results) >= limit,
                query_time_ms=elapsed_ms,
            ),
        )

    def find_callees(
        self,
        symbol_name: str,
        depth: int = 1,
        limit: int = 20,
        file_filter: str | None = None,
        include_external: bool = True,
    ) -> CallGraphResponse:
        """Find all callees of a symbol with transitive depth."""
        start_time = time.time()
        clamped_depth = min(max(depth, 1), 5)

        resolved = self._resolver.resolve(symbol_name)
        if not resolved:
            return self._symbol_not_found(symbol_name)

        results: list[CallGraphItem] = []
        visited: set[str] = set()
        queue: deque[tuple[int, int]] = deque()

        for sym in resolved:
            queue.append((sym.id, 0))

        while queue and len(results) < limit:
            symbol_id, current_depth = queue.popleft()
            if current_depth >= clamped_depth:
                continue

            callees = self._find_callees_db(symbol_id, "calls", limit - len(results))

            for callee in callees:
                key = f"{callee['name']}:{callee['call_line']}"
                if key in visited:
                    continue
                visited.add(key)

                if not include_external and not callee["file_path"]:
                    continue

                item = CallGraphItem(
                    symbol=callee["name"],
                    qualified_name=callee["name"],
                    kind=callee["kind"] or "unknown",
                    file_path=callee["file_path"] or "(external)",
                    definition_line=callee["def_line"] or 0,
                    call_site_line=callee["call_line"],
                    depth_level=current_depth + 1,
                )

                if file_filter and item.file_path != "(external)" and not self._match_filter(item.file_path, file_filter):
                    continue

                results.append(item)

                if callee["file_path"] and current_depth + 1 < clamped_depth:
                    callee_resolved = self._resolver.resolve(callee["name"])
                    for cr in callee_resolved:
                        if cr.file_path == callee["file_path"]:
                            queue.append((cr.id, current_depth + 1))
                            break

        elapsed_ms = int((time.time() - start_time) * 1000)
        return CallGraphResponse(
            symbol=symbol_name,
            resolved_to=[ResolvedTo(id=s.id, file=s.file_path, line=s.line, kind=s.kind) for s in resolved],
            results=results,
            metadata=CallGraphMetadata(
                total_count=len(results),
                depth_searched=clamped_depth,
                truncated=len(results) >= limit,
                query_time_ms=elapsed_ms,
            ),
        )

    def _find_callers_db(self, symbol_name: str, kind: str, limit: int) -> list[dict]:
        cur = self._conn.execute(
            """SELECT s.name, s.kind, s.file_path, s.start_line as def_line,
                      r.line as call_line, s.parent_symbol as parameters,
                      s.visibility as is_async, s.id
               FROM relationships r
               JOIN symbols s ON s.id = r.source_symbol_id
               WHERE r.target_symbol = ? AND r.kind = ?
               ORDER BY s.file_path, r.line
               LIMIT ?""",
            (symbol_name, kind, limit),
        )
        cols = ["name", "kind", "file_path", "def_line", "call_line", "parameters", "is_async", "id"]
        return [dict(zip(cols, row)) for row in cur.fetchall()]

    def _find_callees_db(self, symbol_id: int, kind: str, limit: int) -> list[dict]:
        cur = self._conn.execute(
            """SELECT r.target_symbol as name, r.line as call_line, r.metadata,
                      ts.kind, ts.file_path, ts.start_line as def_line
               FROM relationships r
               LEFT JOIN symbols ts ON ts.id = r.target_symbol_id
               WHERE r.source_symbol_id = ? AND r.kind = ?
               ORDER BY r.line
               LIMIT ?""",
            (symbol_id, kind, limit),
        )
        cols = ["name", "call_line", "metadata", "kind", "file_path", "def_line"]
        return [dict(zip(cols, row)) for row in cur.fetchall()]

    def _symbol_not_found(self, symbol_name: str) -> CallGraphResponse:
        return CallGraphResponse(
            symbol=symbol_name,
            metadata=CallGraphMetadata(query_time_ms=0),
        )

    @staticmethod
    def _match_filter(file_path: str, filter_str: str) -> bool:
        if "*" in filter_str:
            pattern = "^" + re.escape(filter_str).replace(r"\*", ".*") + "$"
            return bool(re.match(pattern, file_path))
        return filter_str in file_path
